compiler/damlc/daml-compiler/src/DA/Daml/Compiler/DataDependencies.hs

```diff
@@ -860,3 +836,81 @@ getClassMethodName (LF.FieldName fieldName) =
 isSuperClassField :: LF.FieldName -> Bool
 isSuperClassField (LF.FieldName fieldName) =
     "s_" `T.isPrefixOf` fieldName
+
+-- | Signature data for a dictionary function.
+data DFunSig
+    = DFunSigNormal -- ^ signature for a normal dictionary function
+        { dfsType :: !LF.Type -- ^ LF type signature for the dictionary function
+        }
+    | DFunSigHasField -- ^ signature for a HasField dictionary function
+        { dfsBinders :: ![(LF.TypeVarName, LF.Kind)] -- ^ foralls
+        , dfsContext :: ![LF.Type] -- ^ constraints
+        , dfsName :: LF.Qualified LF.TypeSynName -- ^ name of type synonym
+        , dfsField :: !T.Text -- ^ first arg (a type level string)
+        , dfsArgs :: ![LF.Type] -- ^ rest of the args
+        }
+
+-- | Break a value type signature down into a dictionary function signature.
+getDFunSig :: (LF.ExprValName, LF.Type) -> Maybe DFunSig
+getDFunSig (valName, valType) = do
+    (dfsBinders, dfsContext, dfsName, args) <- go valType
+    if isHasField dfsName
+        then do
+            (LF.TUnit : dfsArgs) <- Just args
+            dfsField <- getFieldArg valName
+            guard (not $ T.null dfsField)
+            Just DFunSigHasField {..}
+        else
+            Just DFunSigNormal { dfsType = valType }
+  where
+    -- | Break a dictionary function type down into a tuple
+    -- (foralls, constraints, synonym name, args).
+    go = \case
+        LF.TForall b rest -> do
+            (bs, cs, name, args) <- go rest
+            Just (b:bs, cs, name, args)
+        (LF.:->) c rest -> do
+            guard (isConstraint c)
+            (bs, cs, name, args) <- go rest
+            Just (bs, c:cs, name, args)
+        LF.TSynApp name args -> do
+            Just ([], [], name, args)
+        _ ->
+            Nothing
+
+    -- | Is this an instance of the HasField typeclass?
+    isHasField :: LF.Qualified LF.TypeSynName -> Bool
+    isHasField LF.Qualified{..} =
+        qualModule == LF.ModuleName ["DA", "Internal", "Record"]
+        && qualObject == LF.TypeSynName ["HasField"]
+
+    -- | Get the first arg of HasField (the name of the field) from
+    -- the name of the binding.
+    getFieldArg :: LF.ExprValName -> Maybe T.Text
+    getFieldArg (LF.ExprValName name) = do
+        name' <- T.stripPrefix "$fHasField\"" name
```

Comment:
I wonder if it might be slightly less hacky to search for a record selector in the body? Don’t really have a strong opinion here so happy to go with this for now.

Reply:
I think either way is fine, but this seems a little bit simpler, and should handle custom `HasField` instances better. (Not that I suggest anybody write those, but you never know.)